python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeInferenceV2Test.java

```diff
@@ -993,6 +994,84 @@ void try_except_list_attributes() {
 
   }
 
+  @Test
+  void inferTypeForQualifiedExpression() {
+    var root = inferTypes("""
+      class A:
+        def foo():
+           ...
+      def f():
+        a = A()
+        a.foo()
+      """);
+
+    var qualifiedExpression = TreeUtils.firstChild(root.statements().statements().get(1), QualifiedExpression.class::isInstance)
+      .map(QualifiedExpression.class::cast)
+      .get();
+
+    Assertions.assertThat(qualifiedExpression)
+      .isNotNull()
+      .extracting(QualifiedExpression::typeV2)
+      .isNotNull();
+
+    var qualifiedExpressionType = qualifiedExpression.typeV2();
+    Assertions.assertThat(qualifiedExpressionType)
+      .isInstanceOf(FunctionType.class)
+      .extracting(PythonType::name)
+      .isEqualTo("foo");
+  }
+
+
+  @Test
+  void inferBuiltinsTypeForQualifiedExpression() {
+    var root = inferTypes("""
+      a = [42]
+      a.append(1)
+      """);
+
+    var qualifiedExpression = TreeUtils.firstChild(root.statements().statements().get(1), QualifiedExpression.class::isInstance)
+      .map(QualifiedExpression.class::cast)
+      .get();
+
+    Assertions.assertThat(qualifiedExpression)
+      .isNotNull()
+      .extracting(QualifiedExpression::typeV2)
+      .isNotNull();
+
+    var qualifiedExpressionType = qualifiedExpression.typeV2();
+    Assertions.assertThat(qualifiedExpressionType)
+      .isInstanceOf(FunctionType.class)
+      .extracting(PythonType::name)
+      .isEqualTo("append");
+  }
+
+  @Test
+  @Disabled("Attribute types resolving")
+  void inferBuiltinsAttributeTypeForQualifiedExpression() {
+    var root = inferTypes("""
+      def f():
+        e = OSError()
+        e.errno
+      """);
+
+    var qualifiedExpression = TreeUtils.firstChild(root.statements().statements().get(0), QualifiedExpression.class::isInstance)
+      .map(QualifiedExpression.class::cast)
+      .get();
+
+    Assertions.assertThat(qualifiedExpression)
+      .isNotNull()
+      .extracting(QualifiedExpression::typeV2)
+      .isNotNull();
+
```

Comment:
Another test that we may want to have is when the attribute is actually overridden:

```
class A:
  def foo(self): ...

a = A()
a.foo = 42
a.foo() # type of a.foo() ?
```

I don't think this would be supported, and I don't think it was supported in the previous model anyway, but it would be good to document (and compare to what we had before if it's different). I see this could be a source of FPs on which we might want to keep an eye.